Add XML/YAML launch file equivalents to Tf2 tutorials

lsy3 opened the pull request:
## Description

Add XML/YAML launch file equivalents for intermediate Tf2 tutorials.

- Add XML and YAML equivalents for all Python launch files
- Update tutorial documentation with tabbed multi-format code examples

Fixes #6006

### Did you use Generative AI?

Partially - Cursor AI Assistant for generating XML/YAML launch file equivalents and tutorial rephrasing. Each generated line reviewed.

### Additional Information

- Launch files are scoped to pages under `Tutorials/Intermediate/Tf2/` directory
- All launch file references in the Tf2 tutorial documentation have been updated

lsy3 committed:
Add XML/YAML launch file equivalents to Tf2 tutorials

- Add XML and YAML equivalents for all Python launch files
- Update tutorial RST files with tabbed multi-format code examples
- Use .xml as default in console commands with inline format comments

Signed-off-by: Luke Sy <sylukewicent@gmail.com>

lsy3 committed:
Fix RST indentation errors in Tf2 tutorial tabbed code blocks

Signed-off-by: Luke Sy <sylukewicent@gmail.com>

lsy3 committed:
Fix Tf2 tutorial launch file docs and line numbers

- Update launch file format references (.xml -> .py)
- Make launch file sections format-agnostic with tab-specific explanations
- Correct YAML literalinclude line numbers in Adding-A-Frame tutorials

Signed-off-by: Luke Sy <sylukewicent@gmail.com>

lsy3 committed:
Merge branch 'rolling' into 6006-intermediate-tutorial-tf2-xml-yml-launch

kscottz commented:
@lsy3 this is great and super helpful. 

I did a 10 minute read through on this PR and I didn't see any immediate problems. However, I also did not run the launch files in question. The PR touches a lot of files but the changes are fairly repetitive. @emersonknapp if you are happy with this I am willing to do a second approval.

kscottz committed:
Merge branch 'rolling' into 6006-intermediate-tutorial-tf2-xml-yml-launch

kscottz commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -159,19 +159,49 @@ Add the following line between the ``'console_scripts':`` brackets:
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Now let's create a launch file for this example.
```
```suggestion
. attention:: Launch files can be written in XML, YAML, or Python format.
  Throughout this tutorial launch files are shown in all three formats using tabs.
  You can choose whichever format you prefer - they are functionally equivalent.
  Wherever you see the file name ``turtle_tf2_fixed_frame_demo_launch`` make 
  sure to use the correct file extension for your launch file type (i.e.
    ``turtle_tf2_fixed_frame_demo_launch.py`` for Python, 
    ``turtle_tf2_fixed_frame_demo_launch.xml`` for XML, and 
    ``turtle_tf2_fixed_frame_demo_launch.yaml`` for YAML.

Now let's create a launch file for this example.
```

kscottz reviewed (approved):
I think this is acceptable as is. I'll go ahead and approve it. 

We probably want to circle back and write a tutorial comparing and contrasting the three launch file formats and discussing why you might choose one over the other.

kscottz commented:
@SuperJappie08 really appreciate your help on #6021. We would really appreciate your help on this PR too. It would be awesome if we could get these into the docs for the holidays.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Py.rst:
```diff
@@ -252,28 +252,81 @@ Add the following line between the ``'console_scripts':`` brackets:
+  .. group-tab:: XML
+
+    .. literalinclude:: launch/py_turtle_tf2_demo_launch.xml
+        :language: xml
```
The file name is missing.

```suggestion
        :language: xml
        :name: turtle_tf2_demo_launch.xml
```

luke-alloy replied:
Added.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Py.rst:
```diff
@@ -252,28 +252,81 @@ Add the following line between the ``'console_scripts':`` brackets:
+  .. group-tab:: YAML
+
+    .. literalinclude:: launch/py_turtle_tf2_demo_launch.yaml
+        :language: yaml
```
The name is present on the Python variant. 
It should match with the other frontends (either all having a name or none having a name).

```suggestion
        :language: yaml
        :name: turtle_tf2_demo_launch.yaml
```

luke-alloy replied:
Done.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Py.rst:
```diff
@@ -252,28 +252,81 @@ Add the following line between the ``'console_scripts':`` brackets:
-    :name: turtle_tf2_demo_launch.py
+.. tabs::
+
+  .. group-tab:: Python
```
I think the other launch tutorials use the order 
1. XML
2. YAML
3. Python

Is there a reason to deviate from this, as using raw `launch` (without front-end, so most Python launchfiles) is being discouraged?

luke-alloy replied:
Fixed.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Py.rst:
```diff
@@ -395,7 +448,7 @@ Now run the launch file that will start the turtlesim simulation node and ``turt
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_py turtle_tf2_demo_launch.py
+    $ ros2 launch learning_tf2_py turtle_tf2_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be within tabs like all the other launch frontend specific sections.

luke-alloy replied:
Done

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Py.rst:
```diff
@@ -252,28 +252,81 @@ Add the following line between the ``'console_scripts':`` brackets:
+
+    .. literalinclude:: launch/py_turtle_tf2_demo_launch.yaml
+        :language: yaml
+        :lines: 4-9
```
```suggestion
        :lines: 4-14
```
This should probably extend over the whole launchfile

luke-alloy replied:
Opting to extend to the remaining part of the launchfile after the "launch" setup

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Cpp.rst:
```diff
@@ -247,26 +247,79 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
 
-.. literalinclude:: launch/turtle_tf2_demo_launch.py
-    :language: python
+.. tabs::
```
I think the other launch tutorials use the order

1. XML
2. YAML
3. Python

Is there a reason to deviate from this, as using raw `launch` (without front-end, so most Python launchfiles) is being discouraged?

kscottz replied:
At this point we're not giving weight to one over the other. Opinions vary on this and I don't think we are in a position to say, "thou shalt do this." What I am more concerned with is giving the user the list of possible options. 

What I would like to do, once these get merged, is to go to the community and say, "Give me your rationale for using one approach over the other two."  We would then distill this info down into another post that lays out the pros and cons for each approach. I think you can make a compelling case for and against each approach. Anyway, that's why I was looking for the second review ASAP. 😉

SuperJappie08 replied:
I agree that nobody (or global body) should enforce any specific launch file type on anyone.

But I remembered these PR #5093 and #5100, where from the `launch` side of things, active efforts are being made to promote people to use actual `launch` front-ends over using the Python back-end directly.

luke-alloy replied:
Done

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Cpp.rst:
```diff
@@ -377,7 +430,7 @@ Now run the launch file that will start the turtlesim simulation node and ``turt
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.py
+    $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs for the different front-ends?

luke-alloy replied:
Done

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Cpp.rst:
```diff
@@ -247,26 +247,79 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
-.. literalinclude:: launch/turtle_tf2_demo_launch.py
-    :language: python
-    :lines: 7-19
+.. tabs::
```
Make sure this matches the same ordering as above. (`XML`, `YAML`, `Python`)

luke-alloy replied:
Done

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Cpp.rst:
```diff
@@ -247,26 +247,79 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
+
+    .. literalinclude:: launch/turtle_tf2_demo_launch.yaml
+        :language: yaml
+        :lines: 4-9
```
```suggestion
        :lines: 4-14
```

This should probably extend over the whole action

luke-alloy replied:
Opted to show the remaining launch file excluding the "setup"

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Py.rst:
```diff
@@ -227,12 +227,26 @@ Add the following line between the ``'console_scripts':`` brackets:
-.. literalinclude:: launch/listener_py_launch.py
-    :language: python
-    :name: turtle_tf2_demo_launch.py
+.. tabs::
```
Same order comment as made before, as it is probably preferred to be `XML` `YAML` `Python`.

luke-alloy replied:
Done

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Py.rst:
```diff
@@ -227,12 +227,26 @@ Add the following line between the ``'console_scripts':`` brackets:
+
+    .. literalinclude:: launch/listener_py_launch.py
+        :language: python
+        :name: turtle_tf2_demo_launch.py
```
The name should probably be added to all variants or none of them.

luke-alloy replied:
Added

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Py.rst:
```diff
@@ -318,7 +332,7 @@ Now you're ready to start your full turtle demo:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_py turtle_tf2_demo_launch.py
+    $ ros2 launch learning_tf2_py turtle_tf2_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

luke-alloy replied:
Done

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/listener_py_launch.yaml:
```diff
@@ -0,0 +1,33 @@
+%YAML 1.2
+---
+launch:
```
Theoretically, most `"..."` could also be written as `...` except for specific cases, which don't apply here.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/py_turtle_tf2_demo_launch.yaml:
```diff
@@ -0,0 +1,15 @@
+%YAML 1.2
```
> Theoretically, most "..." could also be written as ... except for specific cases, which don't apply here.

The same as the other YAML launchfile, I'm unsure what the general recommendation is.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/turtle_tf2_demo_launch.yaml:
```diff
@@ -0,0 +1,15 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/turtle_tf2_sensor_message_launch.yaml:
```diff
@@ -0,0 +1,31 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/turtle_tf2_fixed_frame_demo_launch.yaml:
```diff
@@ -0,0 +1,10 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/turtle_tf2_dynamic_frame_demo_launch.yaml:
```diff
@@ -0,0 +1,13 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/start_tf2_debug_demo_launch.yaml:
```diff
@@ -0,0 +1,34 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/py_turtle_tf2_fixed_frame_demo_launch.yaml:
```diff
@@ -0,0 +1,10 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/py_turtle_tf2_dynamic_frame_demo_launch.yaml:
```diff
@@ -0,0 +1,13 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/launch/listener_cpp_launch.yaml:
```diff
@@ -0,0 +1,33 @@
+%YAML 1.2
```
Same thing about `"..."` as other YAML launch files

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Cpp.rst:
```diff
@@ -366,7 +380,7 @@ Now you're ready to start your full turtle demo:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.py
+    $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -669,11 +683,11 @@ Open a new terminal, navigate to the root of your workspace, and source the setu
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_py turtle_tf2_sensor_message_launch.py
+    $ ros2 launch learning_tf2_py turtle_tf2_sensor_message_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Time-Travel-With-Tf2-Cpp.rst:
```diff
@@ -112,7 +112,7 @@ Build the package then let's run the simulation again, this time with the advanc
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.py
+    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Time-Travel-With-Tf2-Cpp.rst:
```diff
@@ -53,7 +53,7 @@ Build the package then let's just give it a try:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.py
+    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Learning-About-Tf2-And-Time-Cpp.rst:
```diff
@@ -110,7 +110,7 @@ You can now build the package and run the launch file.
 .. code-block:: console
 
-   $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.py
+   $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Debugging-Tf2-Problems.rst:
```diff
@@ -203,7 +217,7 @@ Stop the demo, build and run:
 .. code-block:: console
 
-   $ ros2 launch turtle_tf2 start_tf2_debug_demo_launch.py
+   $ ros2 launch turtle_tf2 start_tf2_debug_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Debugging-Tf2-Problems.rst:
```diff
@@ -155,7 +169,7 @@ And now stop the running demo, build it, and run it again:
 .. code-block:: console
 
-   $ ros2 launch learning_tf2_cpp start_tf2_debug_demo_launch.py
+   $ ros2 launch learning_tf2_cpp start_tf2_debug_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Debugging-Tf2-Problems.rst:
```diff
@@ -70,18 +70,32 @@ to
 .. code-block:: console
 
-   $ ros2 launch learning_tf2_cpp start_tf2_debug_demo_launch.py
+   $ ros2 launch learning_tf2_cpp start_tf2_debug_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -485,7 +529,7 @@ Now you can start the dynamic frame demo:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_py turtle_tf2_dynamic_frame_demo_launch.py
+    $ ros2 launch learning_tf2_py turtle_tf2_dynamic_frame_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -268,7 +298,7 @@ One way is to pass the ``target_frame`` argument to the launch file directly fro
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_py turtle_tf2_fixed_frame_demo_launch.py target_frame:=carrot1
+    $ ros2 launch learning_tf2_py turtle_tf2_fixed_frame_demo_launch.xml target_frame:=carrot1 # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -253,7 +283,7 @@ Now you can start the turtle broadcaster demo:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_py turtle_tf2_fixed_frame_demo_launch.py
+    $ ros2 launch learning_tf2_py turtle_tf2_fixed_frame_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -528,7 +572,7 @@ Now you can start the dynamic frame demo:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_dynamic_frame_demo_launch.py
+    $ ros2 launch learning_tf2_cpp turtle_tf2_dynamic_frame_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

kscottz replied:
I considered this and I came down on that it is probably fine for one liners. My general metric here is, "Is this better than what we presently have?" The answer in this case is yes.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -290,7 +320,7 @@ One way is to pass the ``target_frame`` argument to the launch file directly fro
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.py target_frame:=carrot1
+    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.xml target_frame:=carrot1 # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -275,7 +305,7 @@ Now you can start the turtle broadcaster demo:
 .. code-block:: console
 
-    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.py
+    $ ros2 launch learning_tf2_cpp turtle_tf2_fixed_frame_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -233,10 +233,24 @@ Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pos
 
-.. literalinclude:: launch/turtle_tf2_sensor_message_launch.py
-  :language: python
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Cpp.rst:
```diff
@@ -278,11 +278,25 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
 
-.. literalinclude:: launch/listener_cpp_launch.py
-  :language: python
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Debugging-Tf2-Problems.rst:
```diff
@@ -70,18 +70,32 @@ to
 
-.. literalinclude:: launch/start_tf2_debug_demo_launch.py
-   :language: python
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -398,10 +428,24 @@ Add the following line between the ``'console_scripts':`` brackets:
-To test this code, create a new launch file ``turtle_tf2_dynamic_frame_demo_launch.py`` in the ``src/learning_tf2_py/launch`` directory and paste the following code:
+To test this code, create a new launch file ``turtle_tf2_dynamic_frame_demo_launch`` with extension ``.py``, ``.xml``, or ``.yaml`` in the ``src/learning_tf2_py/launch`` directory and paste the following code:
+
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -159,19 +159,49 @@ Add the following line between the ``'console_scripts':`` brackets:
-.. literalinclude:: launch/py_turtle_tf2_fixed_frame_demo_launch.py
-    :language: python
-    :lines: 14-18
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -159,19 +159,49 @@ Add the following line between the ``'console_scripts':`` brackets:
-.. literalinclude:: launch/py_turtle_tf2_fixed_frame_demo_launch.py
-    :name: turtle_tf2_fixed_frame_demo_launch.py
-    :language: python
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -442,10 +472,24 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
-To test this code, create a new launch file ``turtle_tf2_dynamic_frame_demo_launch.py`` in the ``src/learning_tf2_cpp/launch`` directory and paste the following code:
+To test this code, create a new launch file ``turtle_tf2_dynamic_frame_demo_launch`` with extension ``.py``, ``.xml``, or ``.yaml`` in the ``src/learning_tf2_cpp/launch`` directory and paste the following code:
+
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -182,18 +182,48 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
-.. literalinclude:: launch/turtle_tf2_fixed_frame_demo_launch.py
-    :language: python
-    :lines: 14-18
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -182,18 +182,48 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
 
-.. literalinclude:: launch/turtle_tf2_fixed_frame_demo_launch.py
-    :language: python
+.. tabs::
```
The order in most other tutorials is `XML`, `YAML`, `Python`, could we make it consistent?

kscottz replied:
That's a little pedantic. We're not implying any particular superiority and we wouldn't enforce it on subsequent PRs.

SuperJappie08 replied:
Same reasoning as other comment https://github.com/ros2/ros2_documentation/pull/6031#discussion_r2613418351

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -159,19 +159,49 @@ Add the following line between the ``'console_scripts':`` brackets:
+  .. group-tab:: Python
+
+    .. literalinclude:: launch/py_turtle_tf2_fixed_frame_demo_launch.py
+        :name: turtle_tf2_fixed_frame_demo_launch.py
```
Can the name be added to all tabs?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -159,19 +159,49 @@ Add the following line between the ``'console_scripts':`` brackets:
+
+    .. literalinclude:: launch/py_turtle_tf2_fixed_frame_demo_launch.xml
+        :language: xml
+        :lines: 3-4
```
Shouldn't this be only line 4?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -398,10 +428,24 @@ Add the following line between the ``'console_scripts':`` brackets:
+
+  .. group-tab:: Python
+
+    .. literalinclude:: launch/py_turtle_tf2_dynamic_frame_demo_launch.py
```
```suggestion
    .. literalinclude:: launch/py_turtle_tf2_dynamic_frame_demo_launch.py
        :language: python
```

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Py.rst:
```diff
@@ -398,10 +428,24 @@ Add the following line between the ``'console_scripts':`` brackets:
+  .. group-tab:: Python
+
+    .. literalinclude:: launch/py_turtle_tf2_dynamic_frame_demo_launch.py
+        :name: turtle_tf2_dynamic_frame_demo_launch.py
```
Can the name be added to all tabs?

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Adding-A-Frame-Cpp.rst:
```diff
@@ -182,18 +182,48 @@ Finally, add the ``install(TARGETS…)`` section so ``ros2 run`` can find your e
+
+    .. literalinclude:: launch/turtle_tf2_fixed_frame_demo_launch.xml
+        :language: xml
+        :lines: 3-4
```
Shouldn't this be line 4 only.

SuperJappie08 commented on source/Tutorials/Intermediate/Tf2/Learning-About-Tf2-And-Time-Cpp.rst:
```diff
@@ -70,7 +70,7 @@ Now build the package and try to run the launch file.
 .. code-block:: console
 
-   $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.py
+   $ ros2 launch learning_tf2_cpp turtle_tf2_demo_launch.xml # .py or .yaml are also acceptable
```
Can this also be tabs per launch file type for consistency between all tutorials?

SuperJappie08 reviewed (changes requested):
It looks good.

Nearly all my comments are minor remarks about consistency with the other tutorials.

EDIT: Sorry for the flooding with comments. I really like consistency.

ahcorde commented:
@lsy3 do you have some time to address @SuperJappie08 comments ?

lsy3 commented:
> @lsy3 do you have some time to address @SuperJappie08 comments ?

I can give it a look within the coming week!

SuperJappie08 commented:
I really went overboard with the comments; I'm still figuring out a _nice_ way to review large PRs.

Any tips are greatly appreciated

kscottz commented:
@SuperJappie08 great question. I think it helps to have the right perspective when reviewing. I've written out my sorta philosophy below.

When it comes to community contributions you shouldn't be asking yourself, "Is this perfect?" you should be asking yourself, "Is this better than what we have now?" If the PR is better than we have now then we should probably merge it. 

The likelihood of the contributor returning to fix the PR drops with every change you request, so you want to be judicious. I would limit big change requests on the part of the author to the things that are absolutely critical (i.e. stuff that's actually broken and requires a big fix). For everything else I try to use the "suggest changes" features. By the time you go back and forth on requesting a change, having the author add it, and then reviewing it again you probably wasted a lot more time than having just suggested a change. I also think this feature is helpful because it sorta signals your expectations without forcing the author to spend more time actually implementing them. In my mind it makes the PR review process a bit less oppositional and much more conversant because you are helping the author get their work merged.

christophebedard commented:
@SuperJappie08 it probably covers a lot of stuff Kat mentioned above, but I think these are pretty good reads: _How to Do Code Reviews Like a Human_: [part 1](https://mtlynch.io/human-code-reviews-1/), [part 2](https://mtlynch.io/human-code-reviews-2/).

christophebedard commented:
> We probably want to circle back and write a tutorial comparing and contrasting the three launch file formats and discussing why you might choose one over the other.

For what it's worth, we currently have these:

1. A one-paragraph comparison here: https://docs.ros.org/en/rolling/How-To-Guides/Migrating-from-ROS1/Migrating-Launch-Files.html#migrating-launch-files
2. A slightly more detailed comparison here: https://docs.ros.org/en/rolling/How-To-Guides/Launch-file-different-formats.html#xml-yaml-or-python-which-should-i-use. I think the wording here should _probably_ be changed to favour XML or YAML over Python a bit more (cc @emersonknapp :stuck_out_tongue:).

lsy3 committed:
Merge branch 'rolling' into 6006-intermediate-tutorial-tf2-xml-yml-launch

lsy3 committed:
Fix launch file ordering and add tabs to broadcaster tutorials

Signed-off-by: Luke Sy <sylukewicent@gmail.com>

lsy3 committed:
Fix launch file ordering and add tabs to listener tutorials

Signed-off-by: Luke Sy <sylukewicent@gmail.com>

lsy3 committed:
Add tabs to remaining tf2 tutorials for launch command consistency

Signed-off-by: Luke Sy <sylukewicent@gmail.com>

lsy3 commented:
Thank you so much for your thorough review, @SuperJappie08! Your eagle eye for detail and focus on consistency is greatly appreciated. I've addressed all your actionable comments! 🎉

## Summary

**✅ Launch File Ordering** - Fixed all tab sections to use XML → YAML → Python order (7 locations across 3 files)

**✅ Code Block Formatting** - Added missing `:name:` directives and removed restrictive `:lines:` options to show complete files

**✅ Tab Organization** - Converted all single code blocks to tabs (11 locations across 8 files) so users can see XML, YAML, and Python variants

**✅ Naming Consistency** - Ensured all code block variants have consistent `:name:` usage

**📝 YAML Quote Style** - After checking the docs, the current files already follow the established pattern (quotes are used consistently in `example_main_launch.yaml`, `different_formats_launch.yaml`, etc.), so no changes needed.

I believe the tutorials now follow the same organizational patterns as the rest of the ROS 2 documentation. Thanks again for helping make these tutorials more consistent! 🙏
